mailpoet/assets/js/src/segments/dynamic/dynamic-segments-filters/fields/woocommerce/purchased-with-attribute.tsx

```diff
@@ -0,0 +1,137 @@
+import { useDispatch, useSelect } from '@wordpress/data';
+import { __ } from '@wordpress/i18n';
+import { useEffect, useRef } from 'react';
+import { Select } from 'common/form/select/select';
+import { ReactSelect } from 'common/form/react-select/react-select';
+import { filter } from 'lodash/fp';
+import { storeName } from '../../../store';
+import {
+  AnyValueTypes,
+  FilterProps,
+  SelectOption,
+  WindowProductAttributes,
+  WooCommerceFormItem,
+} from '../../../types';
+
+export function validatePurchasedWithAttribute(
+  formItems: WooCommerceFormItem,
+): boolean {
+  const purchasedProductWithAttributeIsInvalid =
+    !formItems.operator ||
+    formItems.attribute_id === undefined ||
+    formItems.attribute_term_id === undefined;
+
+  return !purchasedProductWithAttributeIsInvalid;
+}
+
+export function PurchasedWithAttributeFields({
+  filterIndex,
+}: FilterProps): JSX.Element {
+  const segment: WooCommerceFormItem = useSelect(
+    (select) => select(storeName).getSegmentFilter(filterIndex),
+    [filterIndex],
+  );
+  const { updateSegmentFilter } = useDispatch(storeName);
+
+  const productAttributes: WindowProductAttributes = useSelect(
+    (select) => select(storeName).getProductAttributes(),
+    [],
+  );
+
+  const productAttributesOptions = Object.values(productAttributes).map(
+    (attribute) => ({
+      value: attribute.id,
+      label: attribute.label,
+    }),
+  );
+
+  const productAttributeTermsOptionsRef = useRef(null);
+
+  useEffect(() => {
+    if (segment.attribute_id === undefined) {
+      productAttributeTermsOptionsRef.current = null;
+      return;
+    }
+
+    productAttributeTermsOptionsRef.current = productAttributes[
+      segment.attribute_id
+    ].terms.map((term) => ({
+      value: term.term_id,
+      label: term.name,
+    }));
+  }, [segment.attribute_id, productAttributes]);
+
+  useEffect(() => {
+    if (
+      segment.operator !== AnyValueTypes.ALL &&
+      segment.operator !== AnyValueTypes.ANY &&
+      segment.operator !== AnyValueTypes.NONE
+    ) {
+      void updateSegmentFilter({ operator: AnyValueTypes.ANY }, filterIndex);
+    }
+  }, [updateSegmentFilter, segment, filterIndex]);
```

Comment:
I've recently read a nice post about cases where `useEffect` might not be the best solution. I think this might be one of those cases ([Adjusting some state when props changes](https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes)). Using the advice from the post could save us a render. But it is not a big deal.

Reply:
Thanks for the link! I'm going to keep this as is for now since it's such a common pattern in this part of the codebase.